tools/find_tap_name/main.go

```diff
@@ -0,0 +1,129 @@
+// Copyright 2019 The Outline Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+  "encoding/binary"
+  "flag"
+  "fmt"
+  "log"
+  "os"
+
+  "golang.org/x/sys/windows/registry"
+)
+
+const (
+  // netAdaptersKeyPath is the registry location of the system's network adapters.
+  netAdaptersKeyPath = `SYSTEM\CurrentControlSet\Control\Class\{4D36E972-E325-11CE-BFC1-08002BE10318}`
+  // netConfigKeyPath is the registry location of the network adapters network configuration.
+  netConfigKeyPath = `SYSTEM\CurrentControlSet\Control\Network\{4D36E972-E325-11CE-BFC1-08002BE10318}`
+)
+
+// findNetworkAdapterName searches the Windows registry for the name of a network adapter with
+// `componentID`. Since there may be more than one network adapter with the same component ID,
+// selects the most recently installed device in the event of a conflict.
+// Returns an empty string and an error if the device name cannot be found.
+func findNetworkAdapterName(componentID string) (string, error) {
+  netAdaptersKey, err := registry.OpenKey(registry.LOCAL_MACHINE, netAdaptersKeyPath, registry.READ)
+  if err != nil {
+    return "", fmt.Errorf("Failed to open the network adapter registry, %v", err)
+  }
+  defer netAdaptersKey.Close()
+
+  // List all network adapters.
+  adapterKeys, err := netAdaptersKey.ReadSubKeyNames(-1)
+  if err != nil {
+    return "", err
+  }
+
+  // Keep track of the most recently installed adapter name.
+  var name string
+  var installTimestamp uint64
+
+  for _, k := range adapterKeys {
+    adapterKey, err := registry.OpenKey(registry.LOCAL_MACHINE, netAdaptersKeyPath + "\\" + k, registry.READ)
```

Comment:
Could the body of this loop be refactored into a `func(...) (name, timestamp, error)`?  I think something like that would be easier to read, with fewer `continue`s and simpler `defer` behavior.  It could also be unit-tested.

Reply:
Done in 401a2a1